### ISSUE TITLE
Unable to compile Fortran script for inference on simple model

### ISSUE THREAD

**Issue by ChaceAshcraft** [NVIDIA/TorchFort#28]
I am trying to create a Fortran script that simply loads a model created via PyTorch, and does inference on it, and I'm running into trouble I haven't been able to figure out how to fix (not experienced in Fortran). 

The script is as follows:
```
program torch_inference
  use, intrinsic :: iso_fortran_env, only: real32, real64
  use torchfort  ! Ensure this points to TorchFort's Fortran module
  implicit none

  ! Declare variables
  integer :: device
  integer :: istat
  integer :: fake_stream  ! On CPU, so ignored
  character(len=100) :: model_name, config_path
  real(real32), allocatable :: input(:, :), output(:, :)
  integer :: i

  ! Define model name and path
  model_name = 'simple_model'
  config_path = '../dnn_models/simple_model_module_torchscript.pt'
  device = -1

  ! Load the model
  istat = torchfort_load_model(model_name, config_path)
  if (istat /= 0) then
     print *, "Failed to load model!"
     stop
  endif

  ! Prepare input data (adjust dimensions as needed)
  allocate(input(10, 1))
  do i = 1, 10
    input(i, 1) = i
  end do

  ! Allocate output data
  allocate(output(1, 1))  ! Adjust size based on model's output

  ! Run inference
  istat = torchfort_inference(model_name, input, output)
  if (istat /= 0) then
     print *, "Inference failed!"
     stop
  endif

  ! Process output (example)
  print *, "Model output: ", output

end program torch_inference
```

I compiled the code in `TorchFort/src/fsrc` to get `torchfort.mod` and `torchfort_m.o` files, which seem to allow me to use TorchFort, but when trying to compile I get:

```
/usr/bin/ld: /tmp/ccvlcgTz.o: in function `MAIN__':
torch_inference.f90:(.text+0x18b): undefined reference to `__torchfort_MOD_torchfort_load_model'
/usr/bin/ld: torch_inference.f90:(.text+0x4c4): undefined reference to `__torchfort_MOD_torchfort_inference_float_2d'
collect2: error: ld returned 1 exit status
```

on Linux, or

```
Undefined symbols for architecture arm64:
  "___torchfort_MOD_torchfort_inference_float_2d", referenced from:
      _MAIN__ in ccoC7lRu.o
  "___torchfort_MOD_torchfort_load_model", referenced from:
      _MAIN__ in ccoC7lRu.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
```

on Mac. I built the code without GPU (to simplify installation) from source. I did notice some errors during build, but they seemed to to with the examples, and so I wasn't sure they were important. 

Any suggestions would be very appreciated. Thanks1

**Comment by romerojosh** [comment]
Hi @ChaceAshcraft!

It looks like you might not be linking in the TorchFort libraries when you compile your program. You will need to add something like `-L${TORCHFORT_INSTALLATION_DIR}/lib -ltorchfort -ltorchfort_fort` to your link line when building your program.

Let me know if that fixes the compilation issue for you.

**Comment by ChaceAshcraft** [comment]
Thanks for you help! I first notice that I don't have a `lib` directory in my TorchFort installation directory, but I do have one in my `build` directory. That has `libtorchfort_fort.so` and `libtorchfort.so` in it. If I point to that directory with the flags you gave, I get 

```
torch_inference.f90:3:7:

    3 |   use torchfort  ! Ensure this points to TorchFort's Fortran module
      |       1
Fatal Error: Cannot open module file ‘torchfort.mod’ for reading at (1): No such file or directory
compilation terminated.
```

Indeed, there is no `torchfort.mod` in that directory. I tried copying `torchfort.mod` to `TorchFort/build/lib` and also pointing to it with the `-I` flag, but I got the original error about the undefined references.

**Comment by romerojosh** [comment]
Can you please share the entire compile and linking line/s you are using? 

I think it would be better for you to provide an installation directory to CMake and call `make install`, which should provide a better directory structure. For example:
```
cmake -DCMAKE_INSTALL_PREFIX=${TORCHFORT_INSTALL_DIR} <other options> ..
```
followed by
```
make -j <nproc> install
```
Afterward, you should have an installed directory structure like:
```
${TORCHFORT_INSTALL_DIR}/lib:
libtorchfort.so libtorchfort_fort.so
${TORCHFORT_INSTALL_DIR}/include:
torchfort.mod
```
Ensure you have lines like `-I ${TORCHFORT_INSTALL_DIR}/include -L ${TORCHFORT_INSTALL_DIR}/lib -ltorchfort -ltorchfort_fort` in your compilation. For good measure, you should also add the library directory to your `LD_LIBRARY_PATH`.

**Comment by ChaceAshcraft** [comment]
My install is as follows:

```
cmake -DCMAKE_INSTALL_PREFIX="/home/myhome/torchfort_work/TorchFort/" \
      -DTORCHFORT_YAML_CPP_ROOT="/home/myhome/torchfort_work/yaml-cpp/" \
      -DTORCHFORT_BUILD_EXAMPLES=1 \
      -DTORCHFORT_ENABLE_GPU=0 \
      -DCMAKE_PREFIX_PATH="`python -c 'import torch;print(torch.utils.cmake_prefix_path)'`" \
      -Dpybind11_DIR="/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/pybind11/share/cmake/pybind11/" \
    ..
make -j install
```

The console output is:
```
A module that was compiled using NumPy 1.x cannot be run in                                                                                        
NumPy 2.1.3 as it may crash. To support both 1.x and 2.x                                                                                           
versions of NumPy, modules must be compiled with NumPy 2.0.                                                                                        
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.                                                                                
                                                                                                                                                   
If you are a user of the module, the easiest solution will be to                                                                                   
downgrade to 'numpy<2' or try to upgrade the affected module.                                                                                      
We expect that some modules will need time to support NumPy 2.                                                                                     
                                                                                                                                                   
Traceback (most recent call last):  File "<string>", line 1, in <module>                                                                           
  File "/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/__init__.py", line 1477, in <module>                            
    from .functional import *  # noqa: F403                                                                                                        
  File "/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/functional.py", line 9, in <module>                             
    import torch.nn.functional as F                                                                                                                
  File "/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/nn/__init__.py", line 1, in <module>                            
    from .modules import *  # noqa: F403                                                                                                           
  File "/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/nn/modules/__init__.py", line 35, in <module>                   
    from .transformer import TransformerEncoder, TransformerDecoder, \                                                                             
  File "/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/nn/modules/transformer.py", line 20, in <module>                
    device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),                                                   
/home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/nn/modules/transformer.py:20: UserWarning: Failed to initialize NumPy: _
ARRAY_API not found (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:84.)                                                             
  device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),                                                     
-- Caffe2: CUDA detected: 11.5                                                                                                                     
-- Caffe2: CUDA nvcc is: /usr/bin/nvcc                                                                                                             
-- Caffe2: CUDA toolkit directory: /usr                                                                                                            
-- Caffe2: Header version is: 11.5                                                                                                                 
-- /usr/lib/x86_64-linux-gnu/libnvrtc.so shorthash is 65f2c18b                                                                                     
-- USE_CUDNN is set to 0. Compiling without cuDNN support                                                                                          
-- USE_CUSPARSELT is set to 0. Compiling without cuSPARSELt support                                                                                
-- Autodetected CUDA architecture(s):  7.5 7.5                                                                                                     
-- Added CUDA NVCC flags for: -gencode;arch=compute_75,code=sm_75                                                                                  
CMake Warning at /home/mhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:22 (message):        static library kineto_LIBRARY-NOTFOUND not found.
Call Stack (most recent call first):                                                                                                                 /home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:127 (append_torchlib_if_found)
  CMakeLists.txt:105 (find_package)                                                                                                                
                                                                                                                                                   
                                                                         
-- Using yaml-cpp library: /usr/local/lib/libyaml-cpp.a                                                                                            
-- Found pybind11: /home/myhome/anaconda3/envs/torchfort/lib/python3.12/site-packages/pybind11/include (found version "2.13.6")
-- Configuring done                                                                                                                                
-- Generating done                                                                                                                                 
-- Build files have been written to: /home/myhome/torchfort_work/TorchFort/build
Scanning dependencies of target torchfort_fort                                                                                                     
Consolidate compiler generated dependencies of target environments            
[  4%] Built target torchfort_fort                                                                                                                 
[  8%] Built target environments                                                                                                                   
Consolidate compiler generated dependencies of target PyEnvironments                                                                               
[ 12%] Built target PyEnvironments                                                                                                                 
Consolidate compiler generated dependencies of target torchfort                                                                                    
[ 78%] Built target torchfort                                                                                                                      
Scanning dependencies of target train_graph                                                                                                        
Scanning dependencies of target train                                                                                                              
Consolidate compiler generated dependencies of target train_cart_pole                                                                              
Scanning dependencies of target train_distributed                                                                                                  
[ 85%] Linking Fortran executable train                                                                                                            
[ 85%] Linking Fortran executable train_distributed                                                                                                
[ 85%] Building Fortran object examples/fortran/simulation/CMakeFiles/train.dir/simulation.f90.o                                                   
[ 87%] Linking CXX executable train_cart_pole                                                                                                      
[ 89%] Building Fortran object examples/fortran/simulation/CMakeFiles/train.dir/train.f90.o                                                        
[ 91%] Linking Fortran executable train                                                                                                            
/usr/bin/ld: ../../../lib/libtorchfort.so: undefined reference to `YAML::LoadFile(std::string const&)'                                             
/usr/bin/ld: ../../../lib/libtorchfort.so: undefined reference to `YAML::detail::node_data::set_scalar(std::string const&)'                        
/usr/bin/ld: ../../../lib/libtorchfort_fort.so: undefined reference to `MPI_Comm_f2c'                                                              
collect2: error: ld returned 1 exit status                                                                                                         
make[2]: *** [examples/fortran/graph/CMakeFiles/train_graph.dir/build.make:100: examples/fortran/graph/train] Error 1                              
make[1]: *** [CMakeFiles/Makefile2:330: examples/fortran/graph/CMakeFiles/train_graph.dir/all] Error 2                                             
make[1]: *** Waiting for unfinished jobs....                                                                                                       
/usr/bin/ld: ../../../lib/libtorchfort.so: undefined reference to `YAML::LoadFile(std::string const&)'                                             
/usr/bin/ld: ../../../lib/libtorchfort.so: undefined reference to `YAML::detail::node_data::set_scalar(std::string const&)'                        
/usr/bin/ld: ../../../lib/libtorchfort_fort.so: undefined reference to `MPI_Comm_f2c'                                                              
collect2: error: ld returned 1 exit status                                                                                                         
make[2]: *** [examples/fortran/simulation/CMakeFiles/train_distributed.dir/build.make:116: examples/fortran/simulation/train_distributed] Error 1  
make[1]: *** [CMakeFiles/Makefile2:303: examples/fortran/simulation/CMakeFiles/train_distributed.dir/all] Error 2                                  
/usr/bin/ld: ../../../lib/libtorchfort.so: undefined reference to `YAML::LoadFile(std::string const&)'                                             
/usr/bin/ld: ../../../lib/libtorchfort.so: undefined reference to `YAML::detail::node_data::set_scalar(std::string const&)'                        
/usr/bin/ld: ../../../lib/libtorchfort_fort.so: undefined reference to `MPI_Comm_f2c'                                                              
collect2: error: ld returned 1 exit status                                                                                                         
make[2]: *** [examples/fortran/simulation/CMakeFiles/train.dir/build.make:116: examples/fortran/simulation/train] Error 1                          
make[1]: *** [CMakeFiles/Makefile2:276: examples/fortran/simulation/CMakeFiles/train.dir/all] Error 2                                              /usr/bin/ld: /tmp/ccz7d37G.ltrans0.ltrans.o: in function `YAML::detail::node& YAML::detail::node_data::convert_to_node<char [13]>(char const (&) [1
3], std::shared_ptr<YAML::detail::memory_holder>)':                                                                                                
/home/myhome/torchfort_work/yaml-cpp/include/yaml-cpp/node/detail/node_ref.h:37: undefined reference to `YAML::detail::node_data::set_scalar(std:
:string const&)'
/usr/bin/ld: /tmp/ccz7d37G.ltrans0.ltrans.o: in function `YAML::Node::Scalar() const':
/home/myhome/torchfort_work/yaml-cpp/include/yaml-cpp/node/impl.h:171: undefined reference to `YAML::detail::node_data::empty_scalar()'
/usr/bin/ld: /tmp/ccz7d37G.ltrans0.ltrans.o: in function `main':
/home/myhome/torchfort_work/TorchFort/examples/cpp/cart_pole/train.cpp:56: undefined reference to `YAML::LoadFile(std::string const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/cpp/cart_pole/CMakeFiles/train_cart_pole.dir/build.make:114: examples/cpp/cart_pole/train_cart_pole] Error 1
make[1]: *** [CMakeFiles/Makefile2:223: examples/cpp/cart_pole/CMakeFiles/train_cart_pole.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

When I navigate to a separate directory and call 

```
gfortran -I $/home/myhome/torchfort_work/TorchFort/include -L /home/myhome/torchfort_work/TorchFort/lib -ltorchfort -ltorchfort_fort torch_inference.f90 -o torch_inference
```

Which gives:
```
f951: Warning: Nonexistent include directory ‘$/home/myhome/torchfort_work/TorchFort/include’ [-Wmissing-include-dirs]
torch_inference.f90:3:7:

    3 |   use torchfort  ! Ensure this points to TorchFort's Fortran module
      |       1
Fatal Error: Cannot open module file ‘torchfort.mod’ for reading at (1): No such file or directory
compilation terminated.
```

Checking again, the TorchFort directory does not have an `include` or `lib` directory, but `TorchFort/build/` has both.  The contents of `TorchFort/build/include/` is `torchfort.mod`, and the contents of `TorchFort/build/lib/` is `libtorchfort_fort.so` and `libtorchfort.so`. 

Changing the above to:

```
gfortran -I /home/myhome/torchfort_work/TorchFort/build/include -L /home/myhome/torchfort_work/TorchFort/build/lib -ltorchfort -ltorchfort_fort torch_inference.f90 -o torch_inference
```

gives

```
/usr/bin/ld: /tmp/ccYKV9gZ.o: in function `MAIN__':
torch_inference.f90:(.text+0x18b): undefined reference to `__torchfort_MOD_torchfort_load_model'
/usr/bin/ld: torch_inference.f90:(.text+0x4c4): undefined reference to `__torchfort_MOD_torchfort_inference_float_2d'
collect2: error: ld returned 1 exit status
```

**Comment by romerojosh** [comment]
Ok, I see a few things here. It seems that your build of our examples is failing, which is why the `make install` doesn't actually create the `lib` and `include` directories in the install directory. You might want to try setting `-DTORCHFORT_BUILD_EXAMPLES=0` for now to skip those, though it would be good to have an environment where those also build correctly.

As to compiling your program, I think `gfortran` is a bit finicky about the order of sections. I think this should work for you:
```
gfortran -I /home/myhome/torchfort_work/TorchFort/build/include  torch_inference.f90 -o torch_inference -L /home/myhome/torchfort_work/TorchFort/build/lib -ltorchfort_fort -ltorchfort
```

**Comment by ChaceAshcraft** [comment]
Thanks again! Okay, I tried adding the `-DTORCHFORT_BUILD_EXAMPLES=0` flag, and It seems like the code built without error. 

However, running the line you suggested gave me the following errors:

```
/usr/bin/ld: /home/myhome/torchfort_work/TorchFort/build/lib/libtorchfort.so: undefined reference to `YAML::LoadFile(std::string const&)'
/usr/bin/ld: /home/myhome/torchfort_work/TorchFort/build/lib/libtorchfort.so: undefined reference to `YAML::detail::node_data::set_scalar(std::string const&)'
/usr/bin/ld: /home/myhome/torchfort_work/TorchFort/build/lib/libtorchfort_fort.so: undefined reference to `MPI_Comm_f2c'
collect2: error: ld returned 1 exit status
```

Could it have something to do with my YAML or MPI libraries?

**Comment by romerojosh** [comment]
Yes, seems like an issue with both actually.

Can you share how you build yaml-cpp? As an example, this is how we build it in our CPU only docker image: https://github.com/NVIDIA/TorchFort/blob/c2f4ce8bfeb4cb55840b4246299332954bfc9afc/docker/Dockerfile_gnu_cpuonly#L31-L40
In particular, I'm wondering if you are setting `-D_GLIBCXX_USE_CXX11_ABI=0` or not? Give that the unknown references involve `std::string`, this seems like a potential source of that issue.

For MPI, what MPI distribution are you using? The `MPI_Comm_f2c` subroutine isn't available in all distributions; notably MPICH/Cray MPICH is missing it. We have a compiler define that will guard the use against it: https://github.com/NVIDIA/TorchFort/blob/c2f4ce8bfeb4cb55840b4246299332954bfc9afc/src/fsrc/torchfort_m.F90#L77-L86
but the current check in TorchFort to set it is not very advanced, it only works for Cray MPICH if you are using their MPI wrapper (`ftn`),  https://github.com/NVIDIA/TorchFort/blob/c2f4ce8bfeb4cb55840b4246299332954bfc9afc/CMakeLists.txt#L223-L225
If you are using MPICH or an MPI distribution without `MPI_Comm_f2c` you can try commenting out the
```
if (CMAKE_Fortran_COMPILER MATCHES "ftn")
endif()
```
lines so that `-DMPICH` is defined in your build and see if that fixes it for you as a quick workaround. I have a more advanced check for this f2c functionality in another library I develop that I will bring over to TorchFort in a PR shortly.

**Comment by ChaceAshcraft** [comment]
I was building `yaml-cpp` following the basic instructions at `https://github.com/jbeder/yaml-cpp`, so I'm was not setting any flags. I did try reinstalling `yaml-cpp` with the new flags, and rebuilding TorchFort with the line you mentioned commented out. 

I was able to compile the executable! I also noticed that I needed to add the TorchFort library to the environment, e.g.

```
export LD_LIBRARY_PATH=/home/myhome/torchfort_work/TorchFort/build/lib:$LD_LIBRARY_PATH
```

but after that, my executable seems to run as well! I still need to work out some issues with the script, but this has been very helpful. If I run into more trouble, I will open a new issue. Thanks!
